File: cti_dashboard_pro/app/web/js/calculations.js
/**
 * Calculation Core for Cooling Tower Performance
 * Includes Psychrometric properties and Merkel Method (CTI ATC 105)
 */

import { psychrometrics, initPsychroEngine } from './psychro-engine.js';
import { merkelKaVL, initMerkelEngine } from './merkel-engine.js';

export const calculations = {
    init: async (psychroLibPath = './data/psychro_f_alt.bin', merkelLibPath = './data/merkel_poly.bin') => {
        await Promise.all([
            initPsychroEngine(psychroLibPath),
            initMerkelEngine(merkelLibPath)
        ]);
    },

    /**
     * Psychrometric properties calculation using the new engine
     */
    getPsychrometricProps: (twb) => {
        try {
            const props = psychrometrics(twb, twb);
            return { ws: props.HR, hs: props.H, pws: props.P };
        } catch (e) {
            return { ws: 0, hs: 0, pws: 0 };
        }
    },

    /**
     * CTI Merkel demand calculation using new merkelKaVL engine
     */
    calculateDemandKaVL: (twb, hwt, cwt, lgRatio) => {
        try {
            const result = merkelKaVL(hwt, cwt, twb, lgRatio);
            if (result && result.valid) {
                return result.kavl;
            }
            return NaN;
        } catch (e) {
            return NaN;
        }
    },

    calculateSupplyKaVL: (lgRatio, constantC, constantM) => {
        return constantC * Math.pow(lgRatio, -constantM);
    },

    /**
     * Find CWT for given WBT and range percentage
     */
    findCWT: (inputs, wbt, rangePercent, flowPercent) => {
        const designRange = inputs.designHWT - inputs.designCWT;
        const actualRange = designRange * rangePercent / 100;
        const actualLG = inputs.lgRatio * (flowPercent / 100);

        const supplyKaVL = calculations.calculateSupplyKaVL(actualLG, inputs.constantC, inputs.constantM);

        let bestCWT = wbt + 1;
        let minDiff = Infinity;
        let lowApproach = 0.01;
        let highApproach = 80.0;
        const tolerance = 1e-7;

        for (let i = 0; i < 100; i++) {
            let midApproach = (lowApproach + highApproach) / 2;
            let cwtGuess = wbt + midApproach;
            let hwt = cwtGuess + actualRange;

            let demandKaVL;
            try {
                demandKaVL = calculations.calculateDemandKaVL(wbt, hwt, cwtGuess, actualLG);
            } catch (e) { demandKaVL = NaN; }
            
            if (isNaN(demandKaVL) || demandKaVL <= 0) {
                lowApproach = midApproach;
                continue;
            }

            let diff = demandKaVL - supplyKaVL;
            let absDiff = Math.abs(diff);

            if (absDiff < minDiff) {
                minDiff = absDiff;
                bestCWT = cwtGuess;
            }

            if (absDiff < tolerance) {
                break;
            }

            if (diff > 0) {
                lowApproach = midApproach;
            } else {
                highApproach = midApproach;
            }
        }

        return bestCWT;
    },

    /**
     * Solves for the predicted CWT (Performance Prediction Off-Design capability)
     * Given WBT, Range, L/G, C, and m, find the CWT that balances Demand/Supply
     */
    solveOffDesignCWT: (wbt, range, lg, constC, constM) => {
        // First get the physical tower's supply capability
        const supplyKavl = calculations.calculateSupplyKaVL(lg, constC, constM);
        if (isNaN(supplyKavl) || supplyKavl <= 0) {
            return null;
        }

        // We use Bisection Method to find the approach where Demand equals Supply
        // The relationship is invariant: HWT = CWT + Range 
        let lowApproach = 0.01;
        let highApproach = 80.0;
        let bestCwt = NaN;
        let bestDiff = Infinity;
        let matchedDemand = NaN;
        const tolerance = 1e-7;
        const maxIters = 100;
        
        for (let i = 0; i < maxIters; i++) {
            let midApproach = (lowApproach + highApproach) / 2;
            let guessCwt = wbt + midApproach;
            let guessHwt = guessCwt + range;

            let demandKavl;
            try {
                demandKavl = calculations.calculateDemandKaVL(wbt, guessHwt, guessCwt, lg);
            } catch (e) {
                demandKavl = NaN;
            }
            
            if (isNaN(demandKavl) || demandKavl <= 0) {
                // Invalid usually means approach is too low (Demand shoots to infinity)
                lowApproach = midApproach;
                continue;
            }

            let diff = demandKavl - supplyKavl;
            let absDiff = Math.abs(diff);

            if (absDiff < bestDiff) {
                bestDiff = absDiff;
                bestCwt = guessCwt;
                matchedDemand = demandKavl;
            }

            if (absDiff < tolerance) {
                break;
            }

            if (diff > 0) {
                // Demand is higher than Supply. Need more driving force -> increase temperatures (approach).
                lowApproach = midApproach;
            } else {
                // Demand is lower than Supply. Need less driving force -> decrease temperatures (approach).
                highApproach = midApproach;
            }
        }

        if (isNaN(bestCwt) || bestDiff > 0.5) return null; // Unsolvable

        return {
            cwt: bestCwt,
            approach: bestCwt - wbt,
            hwt: bestCwt + range,
            demandKavl: matchedDemand,
            supplyKavl: supplyKavl
        };
    }
};

File: cti_dashboard_pro/app/backend/core/calculations.py
import math
from psychro_engine import psychrometrics, init_psychro_engine
from merkel_engine import merkel_kavl, init_merkel_engine

def init(psychro_lib_path="data/psychro_f_alt.bin", merkel_lib_path="data/merkel_poly.bin"):
    init_psychro_engine(psychro_lib_path)
    init_merkel_engine(merkel_lib_path)

def get_psychrometric_props(twb):
    try:
        props = psychrometrics(twb, twb)
        return {"ws": props["HR"], "hs": props["H"], "pws": props["P"]}
    except Exception:
        return {"ws": 0, "hs": 0, "pws": 0}

def calculate_demand_kavl(twb, hwt, cwt, lg_ratio):
    try:
        result = merkel_kavl(hwt, cwt, twb, lg_ratio)
        if result and result.get("valid"):
            return result["kavl"]
        return float('nan')
    except Exception:
        return float('nan')

def calculate_supply_kavl(lg_ratio, constant_c, constant_m):
    return constant_c * math.pow(lg_ratio, -constant_m)

def find_cwt(inputs, wbt, range_percent, flow_percent):
    design_range = inputs["designHWT"] - inputs["designCWT"]
    actual_range = design_range * range_percent / 100.0
    actual_lg = inputs["lgRatio"] * (flow_percent / 100.0)

    supply_kavl = calculate_supply_kavl(actual_lg, inputs["constantC"], inputs["constantM"])

    best_cwt = wbt + 1
    min_diff = float('inf')
    low_approach = 0.01
    high_approach = 80.0
    tolerance = 1e-7

    for _ in range(100):
        mid_approach = (low_approach + high_approach) / 2.0
        cwt_guess = wbt + mid_approach
        hwt = cwt_guess + actual_range

        try:
            demand_kavl = calculate_demand_kavl(wbt, hwt, cwt_guess, actual_lg)
        except Exception:
            demand_kavl = float('nan')

        if math.isnan(demand_kavl) or demand_kavl <= 0:
            low_approach = mid_approach
            continue

        diff = demand_kavl - supply_kavl
        abs_diff = abs(diff)

        if abs_diff < min_diff:
            min_diff = abs_diff
            best_cwt = cwt_guess

        if abs_diff < tolerance:
            break

        if diff > 0:
            low_approach = mid_approach
        else:
            high_approach = mid_approach

    return best_cwt

def solve_off_design_cwt(wbt, range_val, lg, const_c, const_m):
    supply_kavl = calculate_supply_kavl(lg, const_c, const_m)
    if math.isnan(supply_kavl) or supply_kavl <= 0:
        return None

    low_approach = 0.01
    high_approach = 80.0
    best_cwt = float('nan')
    best_diff = float('inf')
    matched_demand = float('nan')
    tolerance = 1e-7
    max_iters = 100

    for _ in range(max_iters):
        mid_approach = (low_approach + high_approach) / 2.0
        guess_cwt = wbt + mid_approach
        guess_hwt = guess_cwt + range_val

        try:
            demand_kavl = calculate_demand_kavl(wbt, guess_hwt, guess_cwt, lg)
        except Exception:
            demand_kavl = float('nan')

        if math.isnan(demand_kavl) or demand_kavl <= 0:
            low_approach = mid_approach
            continue

        diff = demand_kavl - supply_kavl
        abs_diff = abs(diff)

        if abs_diff < best_diff:
            best_diff = abs_diff
            best_cwt = guess_cwt
            matched_demand = demand_kavl

        if abs_diff < tolerance:
            break

        if diff > 0:
            low_approach = mid_approach
        else:
            high_approach = mid_approach

    if math.isnan(best_cwt) or best_diff > 0.5:
        return None

    return {
        "cwt": best_cwt,
        "approach": best_cwt - wbt,
        "hwt": best_cwt + range_val,
        "demandKavl": matched_demand,
        "supplyKavl": supply_kavl
    }
